init.lua

```diff
@@ -1,116 +1,540 @@
-	privs = {watch=true},
-	func = function(name, param)
-		unwatching(name)
+	-- check for beerchat mute/ignore
+	local meta = get_player_by_name(name_watcher):get_meta()
+	if 'true' == meta:get_string('beerchat:muted:' .. name_target) then
+		return false
+	end
+
+	return true
+end -- is_permited_to_invite
+
+
+-- called by the accept command '/smy'
+local function accept_invite(name_watcher)
+	local name_target = invites[name_watcher]
+	if not name_target then
+		return true, 'There is no invite for you. Maybe it timed-out.'
 	end
+
+	attach(name_watcher, name_target)
+	invites[name_watcher] = nil
+	invited[name_watcher] = name_target
+	chat(name_target, '"' .. name_watcher .. '" is now attached to you.')
+	return true, 'OK, you have been attached to "' .. name_target .. '". To disable type /'
+		.. sm.command_detach
+
+end -- accept_invite
+
+
+-- called by the deny command '/smn'
+local function decline_invite(name_watcher)
+	if not invites[name_watcher] then
+		return true, 'There is no invite for you. Maybe it timed-out.'
+	end
+
+	chat(invites[name_watcher], '"' .. name_watcher .. '" declined the invite.')
+	invites[name_watcher] = nil
+	return true, 'OK, declined invite.'
+end -- decline_invite
+
+
+local function on_joinplayer(watcher)
+	local state = original_state_get(watcher)
+	if not state then return end
+
+	-- attempt to move to original state after log-off
+	-- during attach or server crash
+	local name_watcher = watcher:get_player_name()
+	original_state[name_watcher] = state
+	player_api.player_attached[name_watcher] = true
+	detach(name_watcher)
+end -- on_joinplayer
+
+
+local function on_leaveplayer(watcher)
+	local name_watcher = watcher:get_player_name()
+	if invites[name_watcher] then
+		-- invitation exists for leaving player
+		chat(invites[name_watcher], 'Invitation to "' .. name_watcher
+			.. '" invalidated because of logout.')
+
+		invites[name_watcher] = nil
+	end
+	-- detach before leaving
+	detach(name_watcher)
+	-- detach any that are watching this user
+	local attached = {}
+	for name, state in pairs(original_state) do
+		if name_watcher == state.target then
+			table.insert(attached, name)
+		end
+	end
+	-- we use separate loop to avoid editing a
+	-- hash while it's being looped
+	for _, name in ipairs(attached) do
+		detach(name)
+	end
+end -- on_leaveplayer
+
+
+-- different servers may want different behaviour, they can
+-- override this function
+function spectator_mode.on_respawnplayer(watcher)
+--	* Called when player is to be respawned
+--	* Called _before_ repositioning of player occurs
+--	* return true in func to disable regular player placement
+	local state = original_state_get(watcher)
+	if not state then return end
+
+	local name_target = state.target
+	local name_watcher = watcher:get_player_name()
+	player_api.player_attached[name_watcher] = true
+	if invited[name_watcher] then
+		detach(name_watcher)
+		invited[name_watcher] = name_target
+	else
+		detach(name_watcher)
+	end
```

Comment:
```suggestion
	detach(name_watcher)
	if invited[name_watcher] then
		invited[name_watcher] = name_target
	end
```

Reply:
have added comments in https://github.com/SwissalpS/spectator_mode/commit/6182a9084b38a1eeeb59f8ad19eb4c9a4fb58fab

detach() resets invited[name_watcher] which then makes it too late to check if this observer was invited or had forced himself on other player.

The difference is mainly that if invited, both get informed about the attaching/detaching operations. Also the privs applied matter too on invited status.